src/drivers/usb/class/usb_class_cdc.c

```diff
@@ -12,100 +14,244 @@
+		iface = cdc_get_interface((struct usb_desc_configuration *) cdc->getconf, i);
+		endpoints = (char *) iface + sizeof (struct usb_desc_interface);
+		endpoints += cdc_skip_func_descs(endpoints);
 
-static void usb_class_cdc_get_conf_hnd(struct usb_request *req, void *arg) {
+		for (j = 0; j < iface->b_num_endpoints; j++) {
+			ep = (struct usb_desc_endpoint *) (endpoints +
+					 j * sizeof (struct usb_desc_endpoint));
+			assert(usb_endp_alloc(dev, ep));
+		}
+	}
+}
+
+static void cdc_get_conf_hnd(struct usb_request *req, void *arg) {
 	struct usb_dev *dev = req->endp->dev;
 	struct usb_class_cdc *cdc = usb2cdcdata(dev);
+	struct usb_desc_interface *iface;
 
-	printk("%s: in\n", __func__);
+	iface = cdc_get_interface((struct usb_desc_configuration *) cdc->getconf, 0);
 
-	usb_dev_generic_fill_iface(dev, cdc->getconf + INTERFACE_DESC_OFFSET);
-	usb_dev_generic_fill_endps(dev, cdc->getconf + ENDPOINT_DESC_OFFSET);
+	/* Check if the first interface is one of Microsoft's pet,
+	 * and if so then try to skip it. */
+	if (is_rndis(iface)) {
+		printk("%s\n", "CDC: RNDIS detected");
+		sysfree(cdc->getconf);
+		cdc_next_conf(dev);
+		return;
+	}
 
-	//pool_free(&hid_getconfs, hid->getconf);
-	sysfree(cdc->getconf);
-	cdc->getconf = NULL;
+	cdc_get_endpoints(dev, cdc);
 
-	get_line_coding(dev);
+	usb_driver_handle(req->endp->dev);
 
 	usb_class_start_handle(dev);
-
-	printk("%s: out\n", __func__);
 }
 
-static int usb_class_cdc_get_conf(struct usb_class *cls, struct usb_dev *dev) {
-	size_t conf_size;
+static void cdc_get_conf_len_hnd(struct usb_request *req, void *arg) {
+	struct usb_dev *dev = req->endp->dev;
 	struct usb_class_cdc *cdc = usb2cdcdata(dev);
 
-	conf_size = sizeof(struct usb_desc_configuration)
-					+ sizeof(struct usb_desc_interface)
-					+ FUNCTIONAL_DESCS_TOTAL_SIZE
-					+ (dev->endp_n - 1) * sizeof(struct usb_desc_endpoint);
-
-	cdc->getconf = sysmalloc(conf_size);
+	usb_endp_control(dev->endpoints[0], cdc_get_conf_content_hnd, NULL,
+		USB_DEV_REQ_TYPE_RD
+			| USB_DEV_REQ_TYPE_STD
+			| USB_DEV_REQ_TYPE_DEV,
+		USB_DEV_REQ_GET_DESC,
+		(USB_DESC_TYPE_CONFIG << 8) | cdc->current_conf,
+		dev->c_config,
+		9,
+		cdc_conf);
+}
 
-	usb_endp_control(dev->endpoints[0], cls->get_conf_hnd, NULL,
+static void cdc_get_conf(struct usb_dev *dev, struct usb_class_cdc *cdc,
+		size_t idx, size_t len, usb_request_notify_hnd_t cb) {
+	usb_endp_control(dev->endpoints[0], cb, NULL,
 		USB_DEV_REQ_TYPE_RD
 			| USB_DEV_REQ_TYPE_STD
 			| USB_DEV_REQ_TYPE_DEV,
 		USB_DEV_REQ_GET_DESC,
-		USB_DESC_TYPE_CONFIG << 8,
+		idx,
 		dev->c_config,
-		conf_size,
+		len,
 		cdc->getconf);
+}
+
+static void cdc_get_conf_content_hnd(struct usb_request *req, void *arg) {
+	struct usb_dev *dev = req->endp->dev;
+	struct usb_class_cdc *cdc = usb2cdcdata(dev);
+	struct usb_desc_configuration *c;
+
+	c = (struct usb_desc_configuration*) cdc_conf;
+
+	cdc->getconf = sysmalloc(c->w_total_length);
+	assert(cdc->getconf);
```

Comment:
Have no idea why assert is placed here :) I use <code>if (!data)</code> in other places (e.g., in usbnet.c)